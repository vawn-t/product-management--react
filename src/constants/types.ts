export enum Currencies {
  Dollar = '$',
  Euro = '€',
  VND = '₫'
}

export enum ProductTypes {
  Phone = 'Phone',
  Tablet = 'Tablet',
  Laptop = 'Laptop'
}

export const PRODUCT_TYPE_LIST: ProductTypes[] = [
  ProductTypes.Phone,
  ProductTypes.Tablet,
  ProductTypes.Laptop
];

export enum FilterOrderOptions {
  Asc = 'asc',
  Desc = 'desc'
}

export const ORDER_OPTIONS = [FilterOrderOptions.Asc, FilterOrderOptions.Desc];

export type ValidateError = {
  name: string;
  type: string;
  price: string;
  imageUrl: string;
};

export const INIT_ERRORS: ValidateError = {
  name: '',
  price: '',
  type: '',
  imageUrl: ''
};
