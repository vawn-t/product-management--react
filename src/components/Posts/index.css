.product-group {
  display: block;
  margin: 0 auto;
}

.no-products {
  line-height: 10rem;
  text-align: center;
}
