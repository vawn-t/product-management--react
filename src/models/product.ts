export interface Product {
  readonly id: string;
  name: string;
  type: string;
  price: number;
  imageUrl: string;
}
