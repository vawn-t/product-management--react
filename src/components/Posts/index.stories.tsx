// Library
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Posts from './index';

// Styles
import '@assets/styles/reset.css';
import '@assets/styles/App.css';
import '@assets/styles/variables.css';
import './index.css';

// Model
import { Product } from '@models/product';

// Constants
import { ProductTypes } from '@constants/index';

export default {
  title: 'Components/Posts',
  component: Posts
} as ComponentMeta<typeof Posts>;

const Template: ComponentStory<typeof Posts> = (args) => <Posts {...args} />;

export const Default = Template.bind({});
const products: Product[] = [
  {
    id: '1',
    name: 'iPhone 13 Promax',
    imageUrl:
      'https://cdn-dgcei.nitrocdn.com/QaFavQVnaqgHtiSsAelwGDKVguOuACXM/assets/static/optimized/rev-991ea96/wp-content/uploads/2022/04/iPhone-13-pro-max.png',
    price: 14000000,
    type: ProductTypes.Phone
  },
  {
    id: '2',
    name: 'MacBook Pro 13 M2 2022',
    imageUrl:
      'https://cdn-dgcei.nitrocdn.com/QaFavQVnaqgHtiSsAelwGDKVguOuACXM/assets/static/optimized/rev-991ea96/wp-content/uploads/2022/06/mac-m2.png',
    price: 35990000,
    type: ProductTypes.Laptop
  }
];
Default.args = {
  products
};
