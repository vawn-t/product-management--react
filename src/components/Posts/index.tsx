// Library
import React, { memo } from 'react';

// Components
import { Card , Title} from '@components/index';

// Styles
import './index.css';

// Constants
import { Currencies, ERROR_MESSAGES } from '@constants/index';

// Model
import { Product } from '@models/product';

// Utils
import { areEqual } from '@utils/areEqual';


export interface PostsProps {
  products: Product[];
  onNavigate: (productId: string) => void;
  onDeleteProduct: (id: string) => void;
}

const Posts: React.FC<PostsProps> = ({
  products,
  onNavigate,
  onDeleteProduct
}) => (
  <div className='product-group'>
    {(products || []).map((product: Product) => (
      <Card
        key={product.id}
        product={product}
        currency={Currencies.VND}
        onNavigate={onNavigate}
        onDeleteProduct={onDeleteProduct}
      />
    ))}

    {!products.length && 
      <div className='no-products'>
        <Title>{ERROR_MESSAGES.NO_PRODUCTS_FOUND}</Title>
      </div>
    }
  </div>
);

export default memo(Posts, areEqual<PostsProps>);
