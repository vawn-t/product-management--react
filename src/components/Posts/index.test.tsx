// Libraries
import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';

// Components
import Posts, { PostsProps } from './index';

// Models
import { Product } from '@models/product';

// Constants
import { ProductTypes } from '@constants/types';

describe('Posts component', () => {
  let container: HTMLElement;

  const mockProducts: Product[] = [
    {
      id: '1',
      imageUrl: '',
      name: 'Product test 1',
      price: 1000000,
      type: ProductTypes.Phone
    },
    {
      id: '2',
      imageUrl: '',
      name: 'Product test 2',
      price: 2000000,
      type: ProductTypes.Laptop
    },
    {
      id: '3',
      imageUrl: '',
      name: 'Product test 3',
      price: 3000000,
      type: ProductTypes.Tablet
    }
  ];

  const defaultProps: PostsProps = {
    products: mockProducts,
    onNavigate: jest.fn(),
    onDeleteProduct: jest.fn()
  };
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultTree = create(<Posts {...defaultProps} />).toJSON();
    expect(defaultTree).toMatchSnapshot();
  });

  test('should render correct products length', () => {
    render(<Posts {...defaultProps} />, { container });

    expect(container.getElementsByClassName('card').length).toBe(
      defaultProps.products.length
    );
  });

  test('should handleOpenProductDetail is called', () => {
    const mockOpenProduct = jest.fn();

    render(<Posts {...defaultProps} onNavigate={mockOpenProduct} />, {
      container
    });

    const button: HTMLButtonElement[] = screen.getAllByText('Edit');

    fireEvent.click(button[0]);

    expect(mockOpenProduct).toHaveBeenCalled();
  });

  test('should onDeleteProduct is called', () => {
    const mockDeleteProduct = jest.fn();

    render(<Posts {...defaultProps} onDeleteProduct={mockDeleteProduct} />, {
      container
    });

    const button: HTMLButtonElement[] = screen.getAllByText('Delete');

    fireEvent.click(button[0]);

    expect(mockDeleteProduct).toHaveBeenCalled();
  });
});
