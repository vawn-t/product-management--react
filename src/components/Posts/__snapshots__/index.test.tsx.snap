// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component should render correctly 1`] = `
<div
  className="product-group"
>
  <div
    className="card"
  >
    <div
      className="card-image"
    >
      <img
        alt="Product test 1"
        className="image"
        src=""
      />
    </div>
    <div
      className="card-body"
    >
      <div
        className="title-wrapper"
      >
        <div
          className="title card-title"
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
              "fontStyle": "inherit",
              "fontWeight": undefined,
              "lineHeight": "1.5rem",
            }
          }
        >
          Product test 1
        </div>
      </div>
      <div
        className="subtitle"
        style={
          Object {
            "color": "var(--dark)",
            "fontSize": "16px",
            "fontStyle": "italic",
            "fontWeight": undefined,
            "lineHeight": "2rem",
          }
        }
      >
        Phone
      </div>
      <div
        className="subtitle"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": 700,
            "lineHeight": "1.5rem",
          }
        }
      >
        1 000 000
        <span>
           
          ₫
        </span>
      </div>
      <div
        className="button-wrapper"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="btn btn-secondary"
          disabled={false}
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="card-image"
    >
      <img
        alt="Product test 2"
        className="image"
        src=""
      />
    </div>
    <div
      className="card-body"
    >
      <div
        className="title-wrapper"
      >
        <div
          className="title card-title"
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
              "fontStyle": "inherit",
              "fontWeight": undefined,
              "lineHeight": "1.5rem",
            }
          }
        >
          Product test 2
        </div>
      </div>
      <div
        className="subtitle"
        style={
          Object {
            "color": "var(--dark)",
            "fontSize": "16px",
            "fontStyle": "italic",
            "fontWeight": undefined,
            "lineHeight": "2rem",
          }
        }
      >
        Laptop
      </div>
      <div
        className="subtitle"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": 700,
            "lineHeight": "1.5rem",
          }
        }
      >
        2 000 000
        <span>
           
          ₫
        </span>
      </div>
      <div
        className="button-wrapper"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="btn btn-secondary"
          disabled={false}
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="card-image"
    >
      <img
        alt="Product test 3"
        className="image"
        src=""
      />
    </div>
    <div
      className="card-body"
    >
      <div
        className="title-wrapper"
      >
        <div
          className="title card-title"
          style={
            Object {
              "color": undefined,
              "fontSize": undefined,
              "fontStyle": "inherit",
              "fontWeight": undefined,
              "lineHeight": "1.5rem",
            }
          }
        >
          Product test 3
        </div>
      </div>
      <div
        className="subtitle"
        style={
          Object {
            "color": "var(--dark)",
            "fontSize": "16px",
            "fontStyle": "italic",
            "fontWeight": undefined,
            "lineHeight": "2rem",
          }
        }
      >
        Tablet
      </div>
      <div
        className="subtitle"
        style={
          Object {
            "color": undefined,
            "fontSize": undefined,
            "fontStyle": "inherit",
            "fontWeight": 700,
            "lineHeight": "1.5rem",
          }
        }
      >
        3 000 000
        <span>
           
          ₫
        </span>
      </div>
      <div
        className="button-wrapper"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          className="btn btn-secondary"
          disabled={false}
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
